OpenRA.Mods.D2k/Traits/MakesNoise.cs

```diff
@@ -0,0 +1,51 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	[Desc("This actor makes noise.")]
+	public class MakesNoiseInfo : ITraitInfo
+	{
+		[Desc("List of terrains on which making noise can get you killed.")]
+		public readonly string[] DangerousTerrain = new[] { "Sand", "Dune", "Spice", "SpiceBlobs" };
+
+		[Desc("How much noise does this actor produce.")]
+		public readonly int NoiseIntensity = 0;
+
+		public object Create(ActorInitializer init) { return new MakesNoise(this); }
+	}
+
+	public class MakesNoise : ITick
+	{
+		readonly MakesNoiseInfo info;
+		readonly HashSet<string> dangerousTerrain;
+
+		public int CurrentNoise { get; private set; }
+
+		public MakesNoise(MakesNoiseInfo info)
+		{
+			this.info = info;
+			dangerousTerrain = new HashSet<string>(info.DangerousTerrain);
+		}
+
+		public void Tick(Actor self)
+		{
+			if (self.IsDead)
+				return;
+
+			var terrainName = self.World.Map.GetTerrainInfo(self.Location).Type;
+			
+			CurrentNoise = dangerousTerrain.Contains(terrainName) ? info.NoiseIntensity : 0;
```

Comment:
This doesn't need to run every tick, but only if the current location has changed.  That makes the difference between an array and HashSet even more negligible.